test/ex_machina/ecto_test.exs

```diff
@@ -44,18 +45,38 @@ defmodule ExMachina.EctoTest do
     }
   end
 
-  test "params_for/2 raises when passed a map" do
-    assert_raise ArgumentError, fn ->
-      TestFactory.params_for(:user_map)
-    end
-  end
-
   test "params_for/2 removes fields with nil values" do
     assert TestFactory.params_for(:user, admin: nil) == %{
-      name: "John Doe"
+      name: "John Doe",
+      articles: [],
     }
   end
 
+  test "params_for/2 recursively strips has_many values" do
+    article = TestFactory.build(:article)
+
+    user_params = TestFactory.params_for(:user, articles: [article])
+
+    assert user_params[:articles] == [%{
+      title: article.title,
+    }]
+  end
+
+  test "params_for/2 recursively strips has_one values" do
+    article = TestFactory.build(:article)
+
+    user_params = TestFactory.params_for(:user, best_article: article)
+
+    assert user_params[:best_article] == %{title: article.title}
+  end
+
+  test "params_for/2 works with maps as has_many values" do
+    article = %{title: "Foobar"}
+    user_params = TestFactory.params_for(:user, articles: [article])
```

Comment:
Thank you for your response @paulcsmith!

Yes, _four-phase_ is great patter! In my original comment I was trying to refer to it (though now, when I re-read it, it's not clear - sorry for that).

I was wondering, what's the difference between those two cases, when the [`article` is _built_](https://github.com/thoughtbot/ex_machina/blob/cbc525ec863e3cc05b6b79050e3d2b76ac1255ee/test/ex_machina/ecto_test.exs#L66), and when the [`article` map is prepared _manually_](https://github.com/thoughtbot/ex_machina/blob/cbc525ec863e3cc05b6b79050e3d2b76ac1255ee/test/ex_machina/ecto_test.exs#L74); I have the feeling, that _semantically_ they are exactly the same and as such I'd thought of them to stay in the realm of _setup_ phase (thus, I was asking about the empty line indication).